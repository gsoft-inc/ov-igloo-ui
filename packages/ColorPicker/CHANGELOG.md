# @igloo-ui/color-picker

## 1.0.0

### Major Changes

- 40eb117: Removed the "brand" prop since we found a better alternative for checking the brand.

## 0.3.0

### Minor Changes

- 4846c59: Wrapped Igloo CSS in @layer to isolate it.

### Patch Changes

- Updated dependencies [4846c59]
  - @igloo-ui/color@0.4.0
  - @igloo-ui/dropdown@1.8.0

## 0.2.4

### Patch Changes

- 1f6d0b3: Added a border to the dropdown component and removed the outline from the date picker since it was no longer needed.
- Updated dependencies [1f6d0b3]
  - @igloo-ui/dropdown@1.7.1

## 0.2.3

### Patch Changes

- cfb5619: Updated hopper token versions and button versions.
- Updated dependencies [cfb5619]
  - @igloo-ui/color@0.3.3
  - @igloo-ui/dropdown@1.6.6

## 0.2.2

### Patch Changes

- 7d36815: Updated hopper token versions and igloo icon versions.
- Updated dependencies [7d36815]
  - @igloo-ui/color@0.3.2
  - @igloo-ui/dropdown@1.6.5

## 0.2.1

### Patch Changes

- a2df55d: Updated to latest Hopper Token dependency
- Updated dependencies [a2df55d]
  - @igloo-ui/color@0.3.1
  - @igloo-ui/dropdown@1.6.4

## 0.2.0

### Minor Changes

- 9739155: Gave ColorPicker the new Wokleap look. Color now uses css variables to update its color. Fixed spacing issues in tooltip. Updated hopper token version.

### Patch Changes

- Updated dependencies [9739155]
  - @igloo-ui/color@0.3.0
  - @igloo-ui/dropdown@1.6.3

## 0.1.0

### Minor Changes

- 08de993: Created the Color Picker component.

### Patch Changes

- Updated dependencies [5402bcc]
  - @igloo-ui/color@0.2.0
