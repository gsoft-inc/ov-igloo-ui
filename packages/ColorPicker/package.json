{
  "name": "@igloo-ui/color-picker",
  "version": "1.0.8",
  "main": "dist/ColorPicker.js",
  "module": "dist/ColorPicker.js",
  "types": "dist/ColorPicker.d.ts",
  "styles": "dist/color-picker.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "@igloo-ui/provider": ">=0.1.5",
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/color": "^0.5.4",
    "@igloo-ui/dropdown": "^1.9.1",
    "@igloo-ui/tokens": "^2.1.0",
    "@hopper-ui/tokens": "^4.3.2",
    "classnames": "^2.3.2",
    "react-aria": "^3.32.1",
    "react-aria-components": "^1.1.1",
    "react-stately": "^3.30.1"
  },
  "devDependencies": {
    "@igloo-ui/provider": "^0.1.5"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
