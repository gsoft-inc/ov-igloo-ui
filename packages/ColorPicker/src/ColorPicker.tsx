import * as React from "react";

import cx from "classnames";
import { Button, type ButtonProps, ListBox } from "react-aria-components";
import { mergeProps, FocusScope } from "react-aria";
import type { Selection } from "react-stately";

import Dropdown from "@igloo-ui/dropdown";
import Color from "@igloo-ui/color";
import Checkmark from "@igloo-ui/icons/dist/Checkmark";
import variables from "@igloo-ui/tokens/dist/base10/tokens.json";

import ColorPickerOption from "./ColorPickerOption";

import "./color-picker.scss";

export interface ColorNameItem {
    id: string;
    value: string;
    hoverColor: string;
    iconColor: string;
}

export const colorNamesIgloo = [
    { 
        id: "dandelion200", 
        value: variables.dandelion200, 
        hoverColor: variables.dandelion200, 
        iconColor: variables.samoyed 
    },
    { 
        id: "dandelion400", 
        value: variables.dandelion400, 
        hoverColor: variables.dandelion400, 
        iconColor: variables.samoyed 
    },
    { 
        id: "creamsicle150", 
        value: variables.creamsicle150, 
        hoverColor: variables.creamsicle150, 
        iconColor: variables.samoyed 
    },
    { 
        id: "creamsicle200", 
        value: variables.creamsicle200, 
        hoverColor: variables.creamsicle200, 
        iconColor: variables.samoyed 
    },
    { 
        id: "coral200", 
        value: variables.coral200, 
        hoverColor: variables.coral200, 
        iconColor: variables.samoyed 
    },
    { 
        id: "coral300", 
        value: variables.coral300, 
        hoverColor: variables.coral300, 
        iconColor: variables.samoyed 
    },
    { 
        id: "strawberryFields200", 
        value: variables.strawberryFields200, 
        hoverColor: variables.strawberryFields200, 
        iconColor: variables.samoyed 
    },
    { 
        id: "coral800", 
        value: variables.coral800, 
        hoverColor: variables.coral800, 
        iconColor: variables.samoyed 
    },
    { 
        id: "seaweed300", 
        value: variables.seaweed300, 
        hoverColor: variables.seaweed300, 
        iconColor: variables.samoyed 
    },
    { 
        id: "sky100", 
        value: variables.sky100, 
        hoverColor: variables.sky100, 
        iconColor: variables.samoyed 
    },
    { 
        id: "sky200", 
        value: variables.sky200, 
        hoverColor: variables.sky200, 
        iconColor: variables.samoyed 
    },
    { 
        id: "electricBlue200", 
        value: variables.electricBlue200, 
        hoverColor: variables.electricBlue200, 
        iconColor: variables.samoyed 
    },
    { 
        id: "electricBlue400", 
        value: variables.electricBlue400, 
        hoverColor: variables.electricBlue400, 
        iconColor: variables.samoyed 
    },
    { 
        id: "electricBlue700", 
        value: variables.electricBlue700, 
        hoverColor: variables.electricBlue700, 
        iconColor: variables.samoyed 
    }
] as const;

export const colorNamesWorkleap = [
    { 
        id: "decorativeOption3", 
        value: "var(--hop-decorative-option3-surface)", 
        hoverColor: "var(--hop-decorative-option3-surface-hover)", 
        iconColor: "var(--hop-decorative-option3-icon)" 
    },
    { 
        id: "decorativeOption6", 
        value: "var(--hop-decorative-option6-surface)", 
        hoverColor: "var(--hop-decorative-option6-surface-hover)", 
        iconColor: "var(--hop-decorative-option6-icon)" 
    },
    { 
        id: "decorativeOption1", 
        value: "var(--hop-decorative-option1-surface)", 
        hoverColor: "var(--hop-decorative-option1-surface-hover)", 
        iconColor: "var(--hop-decorative-option1-icon)" 
    },
    { 
        id: "decorativeOption8", 
        value: "var(--hop-decorative-option8-surface)", 
        hoverColor: "var(--hop-decorative-option8-surface-hover)", 
        iconColor: "var(--hop-decorative-option8-icon)" 
    },
    { 
        id: "decorativeOption5", 
        value: "var(--hop-decorative-option5-surface)", 
        hoverColor: "var(--hop-decorative-option5-surface-hover)", 
        iconColor: "var(--hop-decorative-option5-icon)" 
    },
    { 
        id: "decorativeOption4", 
        value: "var(--hop-decorative-option4-surface)", 
        hoverColor: "var(--hop-decorative-option4-surface-hover)", 
        iconColor: "var(--hop-decorative-option4-icon)" 
    },
    { 
        id: "decorativeOption2", 
        value: "var(--hop-decorative-option2-surface)", 
        hoverColor: "var(--hop-decorative-option2-surface-hover)", 
        iconColor: "var(--hop-decorative-option2-icon)" 
    },
    { 
        id: "decorativeOption7", 
        value: "var(--hop-decorative-option7-surface)", 
        hoverColor: "var(--hop-decorative-option7-surface-hover)", 
        iconColor: "var(--hop-decorative-option7-icon)" 
    }
] as const;

export type ColorNameIgloo = typeof colorNamesIgloo[number]["id"];
export type ColorNameWL = typeof colorNamesWorkleap[number]["id"];

export type ColorName = ColorNameWL | ColorNameIgloo;

export interface ColorPickerProps {
    /** Add a specific class to the color picker */
    className?: string;
    /** Add a data-test tag for automated tests */
    dataTest?: string;
    /** Disables the color picker */
    disabled?: boolean;
    /** Callback when a color is selected */
    onSelect?: (color: ColorName) => void;
    /** The selected color */
    selectedColor?: ColorName;
}

const ColorPicker: React.FunctionComponent<ColorPickerProps> = ({
    className,
    dataTest,
    disabled = false,
    onSelect,
    selectedColor: propSelectedColor
}: ColorPickerProps) => {
    const getBrand = (): string => {
        return document.documentElement.getAttribute("data-brand") ?? "igloo";
    };

    const brand = getBrand();

    const getDefaultColor = (): ColorName => {
        return brand === "workleap" ? "decorativeOption3" : "dandelion200";
    };

    const selectedColor = propSelectedColor ?? getDefaultColor();

    let colorNames: ReadonlyArray<Readonly<ColorNameItem>>;

    if (brand === "workleap") {
        colorNames = colorNamesWorkleap;
    } else {
        colorNames = colorNamesIgloo;
    }

    const buttonColorSize = brand === "workleap" ? "medium" : "large";
    const listColorSize = brand === "workleap" ? "large" : "xlarge";
    
    const classes = cx("ids-color-picker", className);
    const [showPicker, setShowPicker] = React.useState(false);
    const selectedColorObject = colorNames.find(color => color.id === selectedColor);
    const selectedColorValue = selectedColorObject?.value;
    const selectedHoverColor = selectedColorObject?.hoverColor;

    const handleSelectionChange = (keys: Selection): void => {
        const color = (keys as Set<ColorName>).keys().next().value;

        if (onSelect) {
            onSelect(color);
        }
        setShowPicker(false);
    };

    return <div className={classes} data-test={dataTest}>
        <Dropdown
            key="colorPickerDropdown"
            role="menu"
            content={
                <FocusScope restoreFocus autoFocus>
                    <div className="ids-color-picker__content">

                        <ListBox 
                            selectionMode="single"
                            selectedKeys={[selectedColor]} 
                            onSelectionChange={handleSelectionChange}
                            items={colorNames}
                            aria-label="Color Options"
                            orientation="horizontal"
                            className="ids-color-picker__list"
                            disallowEmptySelection
                            shouldFocusWrap
                        >
                            {item => {
                                return (
                                    <ColorPickerOption
                                        key={item.id}
                                        id={item.id}
                                        color={item.value}
                                        hoverColor={item.hoverColor}
                                        iconColor={item.iconColor}
                                        icon={<Checkmark size="small" />}
                                        size={listColorSize}
                                    />
                                );
                            }}
                        </ListBox>
                    </div>
                </FocusScope>
            }
            isOpen={showPicker}
            className="ids-action-menu__dropdown"
            onClose={() => setShowPicker(false)}
            renderReference={(
                refProps: React.HTMLProps<HTMLButtonElement>
            ) => {
                const mergedProps = mergeProps({ ...refProps,
                    "data-color": selectedColor }, {
                    onPress: () => {
                        setShowPicker(!showPicker);
                    },
                    className: "ids-color-picker__button",
                    isDisabled: disabled,
                    style: { "--ids-picker-selected-color-hover": selectedHoverColor } as React.CSSProperties
                }) as ButtonProps;

                return <Button
                    {...mergedProps}
                >
                    <Color color={selectedColorValue} size={buttonColorSize} />
                </Button>;
            }
            }
        />
    </div>;
};

export default ColorPicker;
