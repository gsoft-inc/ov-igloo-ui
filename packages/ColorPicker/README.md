# ColorPicker

Color pickers allow users to assign a color to a given element.

<Example is="custom" />

<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/color-picker` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/color-picker
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/color-picker
```

## Usage

Then to use the component in your code just import it!

```jsx
import ColorPicker from '@igloo-ui/color-picker';

const [selectedColor, setSelectedColor] = React.useState("dandelion200");

<ColorPicker 
    onSelect={
        (color) => {
            setSelectedColor(color);
        }          
    } 
    selectedColor={selectedColor} 
/>
```
