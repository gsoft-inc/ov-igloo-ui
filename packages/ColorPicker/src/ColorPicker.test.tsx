/**
  * @jest-environment jsdom
*/
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import MockDropdown from '@igloo-ui/dropdown/src/__mocks__/Dropdown.mock';

import ColorPicker, { ColorName, ColorPickerProps, colorNames } from './ColorPicker';

jest.mock('@igloo-ui/dropdown', () => ({
  __esModule: true,
  default: jest.fn(MockDropdown),
}));

const setup = (props: Partial<ColorPickerProps> = {}) => {
  const defaultProps: ColorPickerProps = {
    dataTest: 'ids-color-picker',
  };
  return render(<ColorPicker {...defaultProps} {...props} />);
};

describe('ColorPicker', () => {
  test('It should render without errors', () => {
  setup();
    const wrapper = screen.getByTestId('ids-color-picker');
    expect(wrapper).toBeInTheDocument();
  });

  test('It should render a snapshot', () => {
    const {asFragment} = setup();
    expect(asFragment()).toMatchSnapshot();
  });

  test('It should show the color picker options when clicked', () => {
    setup();
    const button = screen.getByRole('button');
    fireEvent.click(button);
    const options = screen.getAllByRole('option');
    expect(options.length).toBe(colorNames.length);
  });

  test('It should call the onSelect callback and update the selected item when a color is selected', async () => {
    let selectedColor = 'dandelion200' as ColorName;
    const onSelect = jest.fn((color: ColorName) => {
      selectedColor = color;
    });
    const {rerender} = setup({ selectedColor: selectedColor, onSelect });
    const button = screen.getByRole('button');
    await userEvent.click(button);
    const option = screen.getAllByRole('option')[2];

    await userEvent.click(option);
    expect(onSelect).toHaveBeenCalledWith("creamsicle150");
    rerender(<ColorPicker onSelect={onSelect} selectedColor={selectedColor} />);
    expect(button).toHaveAttribute("data-color", "creamsicle150");
  });

  test('It should not show the color picker options when disabled', () => {
    setup({ disabled: true });
    const button = screen.getByRole('button');
    userEvent.click(button);
    const options = screen.queryAllByRole('option');
    expect(options.length).toBe(0);
  });
});
