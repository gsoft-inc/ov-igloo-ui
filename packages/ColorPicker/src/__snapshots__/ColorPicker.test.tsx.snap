// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker It should render a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-color-picker"
    data-test="ids-color-picker"
  >
    <button
      class="ids-color-picker__button"
      data-color="dandelion200"
      type="button"
    >
      <div
        class="ids-color ids-color--large"
        style="background-color: rgb(253, 222, 136);"
      />
    </button>
  </div>
</DocumentFragment>
`;
